spring-integration-mongodb/src/test/java/org/springframework/integration/mongodb/dsl/MongoDbTests.java

```diff
@@ -0,0 +1,261 @@
+		mongoTemplate.save(createPerson(3, "Oleg", "Zhurakousky"), TEST_COLLECTION);
+		mongoTemplate.save(createPerson(4, "Xavi", "Padro"), TEST_COLLECTION);
+	}
+
+	@Configuration
+	@EnableIntegration
+	public static class ContextConfiguration {
+
+		@Bean
+		public IntegrationFlow gatewaySingleQueryFlow() {
+			return f -> f
+					.handle(queryOutboundGateway("{firstName: 'Xavi'}", true))
+					.channel(getResultChannel());
+		}
+
+		@Bean
+		public IntegrationFlow gatewaySingleQueryWithTemplateFlow() {
+			return f -> f
+					.handle(queryOutboundGatewayWithTemplate("{firstName: 'Xavi'}", true))
+					.channel(getResultChannel());
+		}
+
+		@Bean
+		public IntegrationFlow gatewaySingleQueryExpressionFlow() {
+			return f -> f
+					.handle(queryExpressionOutboundGateway(true))
+					.channel(getResultChannel());
+		}
+
+		@Bean
+		public IntegrationFlow gatewayQueryExpressionFlow() {
+			return f -> f
+					.handle(queryExpressionOutboundGateway(false))
+					.channel(getResultChannel());
+		}
+
+		@Bean
+		public IntegrationFlow gatewayQueryExpressionLimitFlow() {
+			return f -> f
+					.handle(queryExpressionOutboundGateway(false, 2))
+					.channel(getResultChannel());
+		}
+
+		@Bean
+		public MessageChannel getResultChannel() {
+			return MessageChannels.queue().get();
+		}
+
+		@Bean
+		public MongoDbFactory mongoDbFactory() {
+			return new SimpleMongoDbFactory(new MongoClient(), "test");
+		}
+
+		@Bean
+		public MongoOperations mongoTemplate() {
+			return new MongoTemplate(mongoDbFactory());
+		}
+
+		private MongoDbOutboundGatewaySpec queryOutboundGateway(String query, boolean expectSingleResult) {
+			return MongoDb.outboundGateway(mongoDbFactory())
+					.query(query)
+					.collectionNameExpression(new LiteralExpression(TEST_COLLECTION))
+					.expectSingleResult(expectSingleResult)
+					.entityClass(PersonDomain.class);
+		}
+
+		private MongoDbOutboundGatewaySpec queryOutboundGatewayWithTemplate(String query, boolean expectSingleResult) {
+			return MongoDb.outboundGateway(mongoTemplate())
+					.query(query)
+					.collectionNameExpression(new LiteralExpression(TEST_COLLECTION))
+					.expectSingleResult(expectSingleResult)
+					.entityClass(PersonDomain.class);
+		}
+
+		private MongoDbOutboundGatewaySpec queryExpressionOutboundGateway(boolean expectSingleResult) {
+			return MongoDb.outboundGateway(mongoDbFactory())
+					.queryExpression("payload.query")
+					.collectionNameExpression(new LiteralExpression(TEST_COLLECTION))
+					.expectSingleResult(expectSingleResult)
+					.entityClass(PersonDomain.class);
+		}
+
+		private MongoDbOutboundGatewaySpec queryExpressionOutboundGateway(boolean expectSingleResult, int maxResults) {
+			return MongoDb.outboundGateway(mongoDbFactory())
+					.queryExpression("payload.query")
+					.collectionNameExpression(new LiteralExpression(TEST_COLLECTION))
+					.expectSingleResult(expectSingleResult)
+					.entityClass(PersonDomain.class)
+					.maxResults(maxResults);
```

Comment:
When we introduce `BasicQuery` support for the `query-expression` (as well as separate `.query(Query)` option for DSL), we don't need this `maxResults()` any more.